### PR TITLE
linux-v4l2: Support for H264 codec

### PR DESCRIPTION
**Pull request by ggodlewski**
### Description
Added support for format "H264" to capture from the linux-v4l2 plugin.
Build on top of kosmisk-dk PR to also support MJPEG https://github.com/obsproject/obs-studio/pull/3046

### Motivation and Context
H264 tends to have less noise than MJPEG
It also can use less USB bandwidth which might be important if you're using more than 1 webcam on save USB bus.

### How Has This Been Tested?
I've selected "Moition-JPEG" from the "Video Format" drop down.
I've switch to "H264" from the "Video Format" drop down.
I've switch back and forth multiple times with different framerates and resolution.

MJPEG was tested with Logitech c920 and Microdia Digital microscope
H264 was tested with Logitech c920

### Types of changes
Enhanced if conditions to support 1 more codec.
Renamed files.

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by kkartaltepe** [comment]
If you are going to include commits from other PR's please link the PRs you depend on clearly in your description.

**Comment by WizardCM** [comment]
This seems to break (default) first initialisation of YUYV on the Logitech C930e.

With this PR:
```
info: v4l2-input: Start capture from 
error: v4l2-input: Unable to open device
error: v4l2-input: Initialization failed
info: User added source 'Video Capture Device (V4L2)' (v4l2_input) to scene 'Scene 2'
info: v4l2-input: /dev/video1 seems to not support video capture
info: v4l2-input: Found device 'Logitech Webcam C930e' at /dev/video0
info: v4l2-input: Found input 'Camera 1' (Index 0)
info: v4l2-controls: setting default for Power Line Frequency to 2
info: v4l2-controls: setting default for Exposure, Auto to 3
info: v4l2-input: Pixelformat: YUYV 4:2:2 (available)
info: v4l2-input: Pixelformat: Motion-JPEG (available)
info: v4l2-input: Pixelformat: RGB3 (Emulated) (unavailable)
info: v4l2-input: Pixelformat: BGR3 (Emulated) (available)
info: v4l2-input: Pixelformat: YU12 (Emulated) (available)
info: v4l2-input: Pixelformat: YV12 (Emulated) (available)
info: v4l2-input: Stepwise and Continuous framesizes are currently hardcoded
info: v4l2-input: Stepwise and Continuous framerates are currently hardcoded
info: v4l2-input: Pixelformat: YUYV 4:2:2 (available)
info: v4l2-input: Pixelformat: Motion-JPEG (available)
info: v4l2-input: Pixelformat: RGB3 (Emulated) (unavailable)
info: v4l2-input: Pixelformat: BGR3 (Emulated) (available)
info: v4l2-input: Pixelformat: YU12 (Emulated) (available)
info: v4l2-input: Pixelformat: YV12 (Emulated) (available)
info: v4l2-input: Stepwise and Continuous framesizes are currently hardcoded
info: v4l2-input: Stepwise and Continuous framerates are currently hardcoded
info: v4l2-input: Stepwise and Continuous framerates are currently hardcoded
info: v4l2-input: Start capture from /dev/video0
info: v4l2-input: Input: 0
info: v4l2-input: Resolution: 1920x1080
info: v4l2-input: Pixelformat: YUYV
info: v4l2-input: Linesize: 3840 Bytes
info: v4l2-input: Framerate: 5.00 fps
error: v4l2-input: Failed to initialize codec decoder
error: v4l2-input: Initialization failed
```
([full log](https://obsproject.com/logs/HCEM9lQbWFVz9KVs))

Without this PR:
```
info: v4l2-input: Start capture from 
error: v4l2-input: Unable to open device
error: v4l2-input: Initialization failed
info: User added source 'Video Capture Device (V4L2)' (v4l2_input) to scene 'Scene 2'
info: v4l2-input: /dev/video1 seems to not support video capture
info: v4l2-input: Found device 'Logitech Webcam C930e' at /dev/video0
info: v4l2-input: Found input 'Camera 1' (Index 0)
info: v4l2-controls: setting default for Power Line Frequency to 2
info: v4l2-controls: setting default for Exposure, Auto to 3
info: v4l2-input: Pixelformat: YUYV 4:2:2 (available)
info: v4l2-input: Pixelformat: Motion-JPEG (available)
info: v4l2-input: Pixelformat: RGB3 (Emulated) (unavailable)
info: v4l2-input: Pixelformat: BGR3 (Emulated) (available)
info: v4l2-input: Pixelformat: YU12 (Emulated) (available)
info: v4l2-input: Pixelformat: YV12 (Emulated) (available)
info: v4l2-input: Stepwise and Continuous framesizes are currently hardcoded
info: v4l2-input: Stepwise and Continuous framerates are currently hardcoded
info: v4l2-input: Pixelformat: YUYV 4:2:2 (available)
info: v4l2-input: Pixelformat: Motion-JPEG (available)
info: v4l2-input: Pixelformat: RGB3 (Emulated) (unavailable)
info: v4l2-input: Pixelformat: BGR3 (Emulated) (available)
info: v4l2-input: Pixelformat: YU12 (Emulated) (available)
info: v4l2-input: Pixelformat: YV12 (Emulated) (available)
info: v4l2-input: Stepwise and Continuous framesizes are currently hardcoded
info: v4l2-input: Stepwise and Continuous framerates are currently hardcoded
info: v4l2-input: Stepwise and Continuous framerates are currently hardcoded
info: v4l2-input: Start capture from /dev/video0
info: v4l2-input: Input: 0
info: v4l2-input: Resolution: 1920x1080
info: v4l2-input: Pixelformat: YUYV
info: v4l2-input: Linesize: 3840 Bytes
info: v4l2-input: Framerate: 5.00 fps
info: v4l2-input: /dev/video0: select timeout set to 1000000us (5x frame periods)
```
([full log](https://obsproject.com/logs/jI2t1zq2j7O7SZXr))

I'm also not seeing H264 in the list.

**Comment by ggodlewski** [comment]
Ok, it should be fixed. Can you try now? It's possible that the same bug exists in #3046.

**Comment by kkartaltepe** [comment]
> info: v4l2-input: Framerate: 5.00 fps

Does 5fps actually work for you on your webcam? Though initialization sometimes works 5fps results in a frozen image on my C920, I dont think the webcam actually supports it though libv4l2 reports it as supported.

**Comment by WizardCM** [comment]
@kkartaltepe At the time, it did - at least on Ubuntu 20.04. Seems it doesn't work in 21.04, I'll have to do more testing.

**Comment by ggodlewski** [comment]
@kkartaltepe I encountered a freeze issue related to udev bug. Maybe that was the reason. It is fixed in #6167

File: plugins/linux-v4l2/v4l2-codec.h
Comment (RytoEX) [inline plugins/linux-v4l2/v4l2-codec.h:29]: "codec decoding" is sort of redundant ("coder decoder decoding").

Perhaps just "decoding"?

File: plugins/linux-v4l2/v4l2-codec.h
Comment (RytoEX) [inline plugins/linux-v4l2/v4l2-codec.h:39]: Just "decoder" is probably fine.

File: plugins/linux-v4l2/v4l2-codec.h
Comment (RytoEX) [inline plugins/linux-v4l2/v4l2-codec.h:53]: v4l2_destroy_codec -> v4l2_destroy_decoder

File: plugins/linux-v4l2/v4l2-codec.h
Comment (RytoEX) [inline plugins/linux-v4l2/v4l2-codec.h:46]: v4l2_init_codec -> v4l2_init_decoder

File: plugins/linux-v4l2/v4l2-codec.h
Comment (RytoEX) [inline plugins/linux-v4l2/v4l2-codec.h:56]: "Decode a jpeg or h264 frame into an obs frame"

**Review by RytoEX — CHANGES_REQUESTED** [review]
I've suggested some function name changes and some changes to the comments. Would appreciate a second opinion from @kkartaltepe as I'm not very familiar with the v4l2 code, so I don't want to accidentally misrepresent what's happening with my naming suggestions.

Please rebase this PR on the master branch and do not use merge commits. Please squash the commits into a single commit.

File: plugins/linux-v4l2/v4l2-codec.c
Comment (kkartaltepe) [inline plugins/linux-v4l2/v4l2-codec.c:61]: pixel format needs to be inferred from the decoder, this looks like it reintroduces breakages for other yuv formats.

File: plugins/linux-v4l2/v4l2-codec.h
Comment (kkartaltepe) [reply plugins/linux-v4l2/v4l2-codec.h:39]: +1

File: plugins/linux-v4l2/v4l2-helpers.h
Comment (kkartaltepe) [inline plugins/linux-v4l2/v4l2-helpers.h:82]: This looks like it reintroduces pixel format bugs and should be removed.

**Comment by RytoEX** [comment]
Marking this as a draft again because it contains changes from #6167. Will undraft once that has been merged.

File: plugins/linux-v4l2/v4l2-decoder.c
Comment (RytoEX) [inline plugins/linux-v4l2/v4l2-decoder.c:31]: Can these both be true at the same time? If not, then use an `else if` and link them together.

File: plugins/linux-v4l2/v4l2-decoder.c
Comment (RytoEX) [inline plugins/linux-v4l2/v4l2-decoder.c:38]: Can these both be true at the same time? If not, then use an `else if` and link them together.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Just a quick review. Someone more familiar with Linux, V4L2, and/or this plugin may need to go over this and test it.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Per our [Commit Guidelines](https://github.com/obsproject/obs-studio/blob/b3870476324e2dc79a9d30371a1199376bcd0ad7/CONTRIBUTING.rst#commit-guidelines), please make these changes:
1. Use a verb for the first word of the commit message subject.

Nitpick: please use "H.264" in the commit message.

Something like this should satisfy both requests:
> linux-v4l2: Add support for H.264

**Comment by ggodlewski** [comment]
@RytoEX I made those changes in the message. Can you mark it as done? I'm afraid other reviewers might consider this PR as unfinished.

**Review by RytoEX — APPROVED** [review]
The changes I've asked for have been made. This should still be reviewed by @WizardCM and @kkartaltepe when they have time.

**Review by kkartaltepe — APPROVED** [review]

**Comment by kkartaltepe** [comment]
Things look like they still work on my machine, changes look good.

**Pull request merged**
